Support OpenGL GTK3 New API

mqcmd196 opened the pull request:
GSoC2024 Project 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented:
I created a sample from your test and it shows just empty window:
```
#ifdef _WIN32
    #define WIN32_LEAN_AND_MEAN 1
    #define NOMINMAX 1
    #include <windows.h>
#endif

#if defined(__APPLE__)
    #include <OpenGL/gl.h>
    #include <OpenGL/glu.h>
#else
    #include <GL/gl.h>
    #include <GL/glu.h>
#endif

#include "opencv2/core.hpp"
#include "opencv2/core/opengl.hpp"
#include "opencv2/core/cuda.hpp"
#include "opencv2/highgui.hpp"

using namespace cv;
using namespace std;

int main(int argc, char* argv[])
{
    const std::string window_name("gl_test_window");
    const Size image_size(800, 600);

    namedWindow(window_name, WINDOW_OPENGL);
    Mat m = Mat(image_size, CV_8UC3, Scalar(255, 0, 0));
    resizeWindow(window_name, image_size);
    setOpenGlContext(window_name);

    setOpenGlDrawCallback(window_name, [](void*){
        glClear(GL_COLOR_BUFFER_BIT);
        glColor3f(1.0f, 0.0f, 0.0f);
        glBegin(GL_QUADS);
        glVertex2f(-0.5f, -0.5f);
        glVertex2f( 0.5f, -0.5f);
        glVertex2f( 0.5f,  0.5f);
        glVertex2f(-0.5f,  0.5f);
        glEnd();
    });

    imshow(window_name, m);

    while (waitKey(30) != 27)
    {
    }

    setOpenGlDrawCallback(window_name, 0);
    destroyAllWindows();
}
```

mqcmd196 commented:
@asmorkalov 
I updated the codes and confirmed that the GL test in `modules/highgui/test/test_gui.cpp` works well. (This test will be removed when merged.) As I mentioned online, GTK3's OpenGL API only supports OpenGL 3.x, and I updated its test code to 3.x.

I'll continue working on supporting current OpenCV's OpenGL example code.

mqcmd196 commented:
And I squashed some commits into a single commit.

asmorkalov commented:
Looks like we are closer to result, but opengl/opengl.cpp still does not work correctly. I do not see error messages any more and just black window.

asmorkalov commented:
In case I add sample with your test code I see the following error:
```
/usr/bin/ld: CMakeFiles/example_opengl_opengl.dir/opengl.cpp.o: unresolved external symbol «epoxy_glDrawArrays»
//usr/lib/x86_64-linux-gnu/libepoxy.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
samples/opengl/CMakeFiles/example_opengl_opengl.dir/build.make:103: recipe for target 'bin/example_opengl_opengl' failed
make[3]: *** [bin/example_opengl_opengl] Error 1
```

The error means that libepoxy is missing somewhere in cmake. I'll take a look today.

mqcmd196 commented:
> Looks like we are closer to result, but opengl/opengl.cpp still does not work correctly. I do not see error messages any more and just black window.

Thank you. I added the example code for OpenGL 3.2 https://github.com/opencv/opencv/pull/25822/commits/c035d931e0527dea862e04525ac911d262f8fb98

asmorkalov commented on modules/highgui/test/test_gui.cpp:
```diff
@@ -253,6 +253,77 @@ TEST(Highgui_GUI, currentUIFramework)
+    EXPECT_NO_THROW(destroyAllWindows());
+    ASSERT_NO_THROW(namedWindow(window_name, WINDOW_OPENGL));
+    ASSERT_NO_THROW(resizeWindow(window_name, image_size));
+    DrawData data;
+    const char *vertex_shader_source =
+            "#version 330 core\n"
+            "layout (location = 0) in vec3 position;\n"
+            "void main() {\n"
+            "   gl_Position = vec4(position, 1.0);\n"
+            "}\n";
+    const char *fragment_shader_source =
+            "#version 330 core\n"
+            "out vec4 color;\n"
+            "void main() {\n"
+            "   color = vec4(1.0, 1.0, 1.0, 1.0); // white\n"
+            "}\n";
+    GLuint vertex_shader = create_shader(vertex_shader_source, GL_VERTEX_SHADER);
+    GLuint fragment_shader = create_shader(fragment_shader_source, GL_FRAGMENT_SHADER);
+    data.program = glCreateProgram();
+    glAttachShader(data.program, vertex_shader);
+    glAttachShader(data.program, fragment_shader);
+    glLinkProgram(data.program);
+    glUseProgram(data.program);
+    GLfloat vertices[] = {
+            0.0f,  0.5f, 0.0f,
+            -0.5f, -0.5f, 0.0f,
+            0.5f, -0.5f, 0.0f
+    };
+    glGenVertexArrays(1, &data.vao);
+    glBindVertexArray(data.vao);
+    glGenBuffers(1, &data.vbo);
+    glBindBuffer(GL_ARRAY_BUFFER, data.vbo);
+    glBufferData(GL_ARRAY_BUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);
+    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(GLfloat), (GLvoid*)0);
+    glEnableVertexAttribArray(0);
+    glBindVertexArray(0);
+    ASSERT_NO_THROW(setOpenGlDrawCallback(window_name, draw, &data));
+    EXPECT_NO_THROW(waitKey(10000));
+    EXPECT_NO_THROW(setOpenGlDrawCallback(window_name, 0));
+    EXPECT_NO_THROW(destroyAllWindows());
```
Please remove the new test. We cannot run it in CI.

mqcmd196 replied:
https://github.com/opencv/opencv/pull/25822/commits/b9776440c889212cc41f3b0b292c4b48d5c53c4a

mqcmd196 commented on modules/highgui/src/window_gtk.cpp:
```diff
@@ -68,9 +65,14 @@
 #ifdef HAVE_OPENGL
+  #ifdef GTK_VERSION3
+    #include <gtk/gtkglarea.h>
+    #include <epoxy/gl.h>
```
Should I remove from CMake too?

mqcmd196 replied:
https://github.com/opencv/opencv/pull/25822/commits/b9b912ebc02e7782f602505bf9e498cbbf10bd5b

cpoerschke commented on samples/opengl/opengl3_2.cpp:
```diff
@@ -0,0 +1,169 @@
+    if (argc < 2)
+    {
+        cout << "Usage: " << argv[0] << " image" << endl;
+        filename = "lena.jpg";
```
similar to #25635 - perhaps choose a different default file name from the https://github.com/opencv/opencv/tree/4.x/samples/data samples e.g.
```suggestion
        filename = "HappyFish.jpg";
```

mqcmd196 replied:
https://github.com/opencv/opencv/pull/25822/commits/174265724cb21160229b6b5591acba3526fbc261

asmorkalov commented on CMakeLists.txt:
```diff
@@ -1464,9 +1464,14 @@ if(WITH_GTK OR HAVE_GTK)
     status(  "      GThread :" HAVE_GTHREAD THEN "YES (ver ${GTHREAD_VERSION})" ELSE NO)
-    status(  "      GtkGlExt:" HAVE_GTKGLEXT THEN "YES (ver ${GTKGLEXT_VERSION})" ELSE NO)
+    if(HAVE_GTK3)
+      status(  "    Epoxy:" HAVE_EPOXY THEN "YES (ver ${EPOXY_VERSION})" ELSE NO)
```
Epoxy is not needed for OpenCV itself any more. Let's drop it.

asmorkalov commented on cmake/OpenCVFindLibsGUI.cmake:
```diff
@@ -62,7 +62,7 @@ endif()
 ocv_clear_vars(HAVE_OPENGL HAVE_QT_OPENGL)
 if(WITH_OPENGL)
-  if(WITH_WIN32UI OR (HAVE_QT AND QT_QTOPENGL_FOUND) OR HAVE_GTKGLEXT)
+  if(WITH_WIN32UI OR (HAVE_QT AND QT_QTOPENGL_FOUND) OR (HAVE_GTK3 AND HAVE_EPOXY) OR (HAVE_GTK AND NOT HAVE_GTK3 AND HAVE_GTKGLEXT))
```
The same here

asmorkalov commented on modules/highgui/cmake/detect_gtk.cmake:
```diff
@@ -26,25 +26,33 @@ if(WITH_GTK)
+        ocv_add_external_target(gtkglext "${GTKGLEXT_INCLUDE_DIRS}" "${GTKGLEXT_LIBRARIES}" "HAVE_GTKGLEXT")
+      endif()
+    endif()
+  if(HAVE_GTK3)
+      ocv_check_modules(EPOXY epoxy)
+      if(HAVE_EPOXY)
+        ocv_add_external_target(epoxy "${EPOXY_INCLUDE_DIRS}" "${EPOXY_LIBRARIES}" "HAVE_EPOXY")
+      endif()
```
And here.

mqcmd196 commented:
Sorry, I forgot to remove them.

mqcmd196 commented on modules/core/include/opencv2/core/opengl.hpp:
```diff
@@ -57,7 +57,7 @@ This section describes OpenGL interoperability.
 To enable OpenGL support, configure OpenCV using CMake with WITH_OPENGL=ON . Currently OpenGL is
 supported only with WIN32, GTK and Qt backends on Windows and Linux (MacOS and Android are not
-supported). For GTK backend gtkglext-1.0 library is required.
+supported). For GTK-2.0 backend gtkglext-1.0 library is required.
```
I'll add the comment about `libepoxy` when I create the new wrapper for OpenGL3.2

asmorkalov commented:
Tested manually with:
- Ubuntu 18.04 + GTK 2
- Ubuntu 18.04 + GTK 3
- Ubuntu 18.04 + QT 5.9
- Ubuntu 24.04 + GTK 3
- Ubuntu 24.04 + QT 5.15
